server/src/test/java/org/opensearch/common/cache/stats/DefaultCacheStatsHolderTests.java

```diff
@@ -127,49 +128,68 @@ public void testCount() throws Exception {
-        List<String> dimensionNames = List.of("dim1", "dim2");
+        List<String> dimensionNames = List.of("A", "B");
         DefaultCacheStatsHolder cacheStatsHolder = new DefaultCacheStatsHolder(dimensionNames, storeName);
 
         // Create stats for the following dimension sets
-        List<List<String>> populatedStats = List.of(List.of("A1", "B1"), List.of("A2", "B2"), List.of("A2", "B3"));
+        List<List<String>> populatedStats = new ArrayList<>();
+        int numAValues = 10;
+        int numBValues = 2;
+        for (int indexA = 0; indexA < numAValues; indexA++) {
+            for (int indexB = 0; indexB < numBValues; indexB++) {
+                populatedStats.add(List.of("A" + indexA, "B" + indexB));
+            }
+        }
         for (List<String> dims : populatedStats) {
             cacheStatsHolder.incrementHits(dims);
         }
 
-        // Remove (A2, B2) and (A1, B1), before re-adding (A2, B2). At the end we should have stats for (A2, B2) but not (A1, B1).
+        // Remove a subset of the dimensions concurrently.
+        // Remove both (A0, B0), and (A0, B1), so we expect the intermediate node for A0 to be null afterwards.
+        // For all the others, remove only the B0 value. Then we expect the intermediate nodes for A1 through A9 to be present
+        // and reflect only the stats for their B1 child.
 
-        Thread[] threads = new Thread[3];
-        CountDownLatch countDownLatch = new CountDownLatch(3);
-        threads[0] = new Thread(() -> {
-            cacheStatsHolder.removeDimensions(List.of("A2", "B2"));
-            countDownLatch.countDown();
-        });
-        threads[1] = new Thread(() -> {
-            cacheStatsHolder.removeDimensions(List.of("A1", "B1"));
-            countDownLatch.countDown();
-        });
-        threads[2] = new Thread(() -> {
-            cacheStatsHolder.incrementMisses(List.of("A2", "B2"));
-            cacheStatsHolder.incrementMisses(List.of("A2", "B3"));
+        Thread[] threads = new Thread[numAValues + 1];
+        CountDownLatch countDownLatch = new CountDownLatch(numAValues + 1);
+        Phaser phaser = new Phaser(numAValues + 2);
+        for (int i = 0; i < numAValues; i++) {
+            int finalI = i;
+            threads[i] = new Thread(() -> {
+                phaser.arriveAndAwaitAdvance();
+                cacheStatsHolder.removeDimensions(List.of("A" + finalI, "B0"));
+                countDownLatch.countDown();
+            });
+        }
+        threads[numAValues] = new Thread(() -> {
+            phaser.arriveAndAwaitAdvance();
+            cacheStatsHolder.removeDimensions(List.of("A0", "B1"));
             countDownLatch.countDown();
         });
         for (Thread thread : threads) {
             thread.start();
-            // Add short sleep to ensure threads start their functions in order (so that incrementing doesn't happen before removal)
-            Thread.sleep(1);
         }
+
+        phaser.arriveAndAwaitAdvance();
```

Comment:
I'm confused about the purpose of this Phaser and CountDownLatch. It will cause all the thread functions to block on their first line of execution, and then this call will unblock all of them simultaneously, at which point they will execute their functions concurrently. The very next line will then just wait for them all to finish because they all count down the latch at the end of execution. This looks logically equivalent to just doing:

```
for (Thread thread : threads) {
    thread.join();
}
```

What am I missing?

Reply:
You're right - I'd thought the phaser might make sure they started closer together rather than just starting them sequentially by 
```
for (Thread thread : threads) {
     thread.start();
}
```
 and then waiting for them to finish. But it shouldn't make a difference. Will remove the phaser/latch and run .join() in a for loop instead.

Reply:
> starting them sequentially

You are starting them sequentially, which essentially just schedules the threads as runnable. They may actually execute in any order despite being started in a particular order, so make sure your test doesn't care in which order they run.